From e03b5f07c18c181f9de0a9d0885b90396e50717f Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Thu, 16 Feb 2023 12:18:30 +0800
Subject: [PATCH] header: linux/module.h: move kallsyms to the end of struct

kernel: compatible modules without kallsyms on KALLSYMS enabled kernel

for better compatible
---
 include/linux/module.h | 24 ++++++++++++------------
 kernel/module-internal.h |  1 +
 kernel/module.c          | 37 +++++++++++++++++++++++++++++++++++++
 3 files changed, 50 insertions(+), 12 deletions(-)

--- a/include/linux/module.h
+++ b/include/linux/module.h
@@ -442,18 +442,6 @@ struct module {
 	struct bug_entry *bug_table;
 #endif
 
-#ifdef CONFIG_KALLSYMS
-	/* Protected by RCU and/or module_mutex: use rcu_dereference() */
-	struct mod_kallsyms __rcu *kallsyms;
-	struct mod_kallsyms core_kallsyms;
-
-	/* Section attributes */
-	struct module_sect_attrs *sect_attrs;
-
-	/* Notes attributes */
-	struct module_notes_attrs *notes_attrs;
-#endif
-
 	/* The command line arguments (may be mangled).  People like
 	   keeping pointers to this stuff */
 	char *args;
@@ -537,6 +525,18 @@ struct module {
 	struct error_injection_entry *ei_funcs;
 	unsigned int num_ei_funcs;
 #endif
+
+#ifdef CONFIG_KALLSYMS
+	/* Protected by RCU and/or module_mutex: use rcu_dereference() */
+	struct mod_kallsyms __rcu *kallsyms;
+	struct mod_kallsyms core_kallsyms;
+
+	/* Section attributes */
+	struct module_sect_attrs *sect_attrs;
+
+	/* Notes attributes */
+	struct module_notes_attrs *notes_attrs;
+#endif
 } ____cacheline_aligned __randomize_layout;
 #ifndef MODULE_ARCH_INIT
 #define MODULE_ARCH_INIT {}
--- a/kernel/module-internal.h
+++ b/kernel/module-internal.h
@@ -22,6 +22,7 @@ struct load_info {
 	bool sig_ok;
 #ifdef CONFIG_KALLSYMS
 	unsigned long mod_kallsyms_init_off;
+	unsigned long mod_no_kallsyms_pat;
 #endif
 	struct {
 		unsigned int sym, str, mod, vers, info, pcpu;
--- a/kernel/module.c
+++ b/kernel/module.c
@@ -2534,6 +2534,12 @@ static void layout_sections(struct modul
 			break;
 		}
 	}
+#ifdef CONFIG_KALLSYMS
+	if (info->mod_no_kallsyms_pat) {
+		info->mod_no_kallsyms_pat = mod->core_layout.size;
+		mod->core_layout.size = debug_align(sizeof(struct module) + mod->core_layout.size);
+	}
+#endif
 
 	pr_debug("Init section allocation order:\n");
 	for (m = 0; m < ARRAY_SIZE(masks); ++m) {
@@ -3177,6 +3183,10 @@ out:
 static void free_copy(struct load_info *info)
 {
 	vfree(info->hdr);
+#ifdef CONFIG_KALLSYMS
+	if (info->mod_no_kallsyms_pat)
+		vfree(info->mod);
+#endif
 }
 
 static int rewrite_section_headers(struct load_info *info, int flags)
@@ -3213,6 +3223,7 @@ static int rewrite_section_headers(struc
 static int setup_load_info(struct load_info *info, int flags)
 {
 	unsigned int i;
+	void *ptr;
 
 	/* Try to find a name early so we can log errors with a module name */
 	info->index.info = find_sec(info, ".modinfo");
@@ -3252,6 +3263,24 @@ static int setup_load_info(struct load_i
 	if (!info->name)
 		info->name = info->mod->name;
 
+#ifdef CONFIG_KALLSYMS
+	info->mod_no_kallsyms_pat = 0;
+	if (sizeof(struct module) > info->sechdrs[info->index.mod].sh_size) {
+		pr_warn("%s: module has no kallsyms\n",
+			info->name);
+		ptr = info->mod;
+		info->mod = __vmalloc(sizeof(struct module), GFP_KERNEL | __GFP_NOWARN);
+		if (!info->mod) {
+			pr_err("%s: alloc module copy failed\n",
+						info->name);
+			return -ENOMEM;
+		}
+		info->mod_no_kallsyms_pat = 1;
+		memset(info->mod, 0, sizeof(struct module));
+		memcpy(info->mod, ptr, info->sechdrs[info->index.mod].sh_size);
+	}
+#endif
+
 	if (flags & MODULE_INIT_IGNORE_MODVERSIONS)
 		info->index.vers = 0; /* Pretend no __versions section! */
 	else
@@ -3478,6 +3507,14 @@ static int move_module(struct module *mo
 
 		if (shdr->sh_type != SHT_NOBITS)
 			memcpy(dest, (void *)shdr->sh_addr, shdr->sh_size);
+
+#ifdef CONFIG_KALLSYMS
+		if (info->mod_no_kallsyms_pat && i == info->index.mod) {
+			dest = mod->core_layout.base + info->mod_no_kallsyms_pat;
+			memcpy(dest, mod, sizeof(struct module));
+		}
+#endif
+
 		/* Update sh_addr to point to copy in image. */
 		shdr->sh_addr = (unsigned long)dest;
 		pr_debug("\t0x%lx %s\n",
